crates/datasources/src/lib.rs

```diff
@@ -17,3 +30,49 @@ pub mod postgres;
 pub mod snowflake;
 pub mod sqlite;
 pub mod sqlserver;
+
+
+pub type DatasourceError = Box<dyn std::error::Error + Send + Sync>;
+#[async_trait]
+pub trait Datasource: Send + Sync {
+    fn name(&self) -> &'static str;
+
+    /// Validate the provided tunnel options.
+    fn validate_tunnel_connections(
+        &self,
+        tunnel_opts: Option<&TunnelOptions>,
+    ) -> Result<(), DatasourceError>;
+
+    /// Validate the provided credentials.
+    fn validate_credentials(
+        &self,
+        creds: Option<CredentialsOptions>,
+    ) -> Result<(), DatasourceError>;
+
+    /// Create a new datasource from the provided table options and credentials.
+    /// CREATE EXTERNAL TABLE foo FROM <name> OPTIONS (...) [CREDENTIALS] (...) [TUNNEL] (...)
+    // TODO: the datasource should have control over it's own CredentialsOptions and TunnelOptions
+    fn table_options_from_stmt(
+        &self,
+        opts: &mut StatementOptions,
+        creds: Option<CredentialsOptions>,
+        tunnel_opts: Option<TunnelOptions>,
+    ) -> Result<TableOptions, DatasourceError>;
+
+
+    async fn create_table_provider(
+        &self,
+        tbl_options: &TableOptions,
+        _tunnel_opts: Option<&TunnelOptions>,
```

Comment:
in my mind, tunnel options (like credentials, really, are a component of the table options and depend on the implementation details of the provider rather than something that's pulled out at this level. I understand that it's not for historical reasons, and maybe can't be yet in the process, but I think it'd be good to:
- move toward that if it's possible
- explain why it's not possible if it's not/neer going to be possible
- leave a note/plan here if it's possible but we want to defer it.

Reply:
I agree that they are a component of the datasource, but disagree that they should be part of the table options. The table options should be a 1-1 mapping of the sql `options (...)`, similar to `credentials` and `tunnel`. Instead of trying to throw everything into the `TableOptions`, i'd prefer that `CredentialsOptions` and `TunnelOptions` are refactored similarly to `TableOptions`. 

Each datasource should optionally provide their own `CredentialsOptions`, and/or `TunnelOptions` variant. I already added a small note in the `Datasource` trait definition expressing this.

Reply:
>  The table options should be a 1-1 mapping of the sql options (...), similar to credentials and tunnel. 

I'm not sure I understand this: the indirection between the statements issued by the users and the table options affords us some extra flexibility and provides the ability for both the implementation and the syntax to be easier to use and develop. By forcing our internal representation to mirror the syntax, all of the table options code, becomes inflexible and we don't actually benefit very much from the indirection. 

If the design means that credentials and tunnel options have to be orthogonal to table options, and we want to handle things like explicit schema orthogonally, then every time we need to add a new feature it sort of seems like we'll have a discussion about 'should this be a new orthogonal configuration category or go in the table options' which doesn't feel productive. 

If authentication (credentials/tunnel) can't be part of the table options then it sort of feels like the only thing in the table options are "location/path" and "format" (if that's not on the provider itself,) and so the options just become this catch all for "everything that isn't top level," which feels like it'll both make the code (and the prospect of implementing a new datasource or adding a feature to an existing datasource) quite hard. 

I might be extrapolating too far along some of these lines, and would definitely be interested in hearing your vision for this, but maybe we should talk about this synchronously sometime soon, but I think we need to have some alignment on how we expect to use these in the future.

Reply:
> If the design means that credentials and tunnel options have to be orthogonal to table options, and we want to handle things like explicit schema orthogonally, then every time we need to add a new feature it sort of seems like we'll have a discussion about 'should this be a new orthogonal configuration category or go in the table options' which doesn't feel productive.

i think there are just a few top level arguments. 

- any datasource can accept a schema (columns). 
- any datasource can require credentials. 
- any datasource can be behind a tunnel.

inversely, 
- not all datasources specify a "format"
- nod all datasources specify a "compression" 

These are all examples of datasource specific arguments. 

The top level ones should be quite limited, and applicable to *any* datasource.

Reply:
While we can imagine "any datasource can accept <...>" as being true in principle, it's absolutely not true in practice:
https://github.com/glaredb/glaredb/blob/main/crates/sqlbuiltins/src/validation.rs#L63-L124

Reply:
After looking at all of this, I'm strongly of the opinion creds/tunnels should be tied to table options.

Passing in creds/tunnels restricts what we're able to do in the future as it relates to the SQL commands themselves (what would happen if I want a multi tunnel hop with aws private link to ssh, or how might I specify backup credentials in the case of rolling creds). These are concerns of the data source, not of the planner. This I think brings up a secondary point of data sources can be hooked up as external tables _or_ external databases, which I think highlights a need for something more along the lines of `DatasourceOptions` and not table specific options.

But at far as avoiding passing in creds/tunnels in directly, I think something like this is reasonable:

```rust
/// Resolve external dependencies for data sources.
pub trait Resolver {
    fn credentials_by_name(&self, name: &str) -> Result<Option<CredentialsOptions>, DatasourceError>;
    fn credentials_by_oid(&self, oid: u32) -> Result<Option<CredentialsOptions>, DatasourceError>;

    fn tunnel_by_name(&self, name: &str) -> Result<Option<TunnelOptions>, DatasourceError>;
    fn tunnel_by_oid(&self, oid: u32) -> Result<Option<TunnelOptions>, DatasourceError>;
}

#[async_trait]
pub trait Datasource: Send + Sync {
    fn name(&self) -> &'static str;

    fn table_options_from_stmt(
        &self,
        opts: &mut StatementOptions,
        resolver: &dyn Resolver,
    ) -> Result<TableOptions, DatasourceError>;


    async fn create_table_provider(
        &self,
        tbl_options: &TableOptions,
        resolver: &dyn Resolver,
    ) -> Result<Arc<dyn TableProvider>, DatasourceError>;
}
```

And elsewhere:

```rust
pub struct CatalogResolver<'a> {
    catalog: &'a SessionCatalog,
}

impl<'a> CatalogResolver<'a> {
    fn credentials_by_name(&self, name: &str) -> Result<Option<CredentialsOptions>, DatasourceError> {...}
    fn credentials_by_oid(&self, oid: u32) -> Result<Option<CredentialsOptions>, DatasourceError> {...}

    fn tunnel_by_name(&self, name: &str) -> Result<Option<TunnelOptions>, DatasourceError> {...}
    fn tunnel_by_oid(&self, oid: u32) -> Result<Option<TunnelOptions>, DatasourceError> {...}
}
```

Reply:
Sounds good, I was also thinking of passing in a resolver to the datasource trait.

Reply:
Is the resolver not just a higher level wrapper around the catalog+session?

Reply:
It is, but provides two benefits:

- `datasources` won't depend on the catalog directly
- If we have a "secure store" eventually we could just abstract over that too